### ISSUE TITLE
Server filter unable to modify response headers for reactive controllers

### ISSUE THREAD

**Issue by lightoze** [micronaut-projects/micronaut-core#3983]
Looks like server filter cannot modify response headers when controller has a signature of `Publisher<MutableHttpResponse<?>>`. As a result, for example HTTP session cannot be created during such controller invocation because no cookie is returned to the client.

I believe this is also causing https://github.com/micronaut-projects/micronaut-security/issues/319

Here is a Gist with simple code to reproduce the problem: https://gist.github.com/lightoze/32f6f43b4a8fd8f3a12c2d288489f07c

Micronaut version: 2.0.1

**Comment by lightoze** [comment]
Here's my temporary workaround:
```java
@Filter("/**")
public class ReactiveResponseFixFilter extends OncePerRequestHttpServerFilter {

    @Override
    public int getOrder() {
        return ServerFilterPhase.RENDERING.before();
    }

    @Override
    protected Publisher<MutableHttpResponse<?>> doFilterOnce(HttpRequest<?> request, ServerFilterChain chain) {
        RouteMatch<?> routeMatch = request.getAttribute(HttpAttributes.ROUTE_MATCH, RouteMatch.class).orElse(null);

        if (routeMatch != null && routeMatch.getReturnType().isReactive() && routeMatch.getReturnType().getFirstTypeVariable()
                .stream().allMatch(arg -> MutableHttpResponse.class.isAssignableFrom(arg.getType()))) {
            return Publishers.map(chain.proceed(request), response -> {
                Object body = response.getBody().orElse(null);
                if (body instanceof Publisher) {
                    Object bodyResult = Flowable.fromPublisher((Publisher<?>) body).blockingSingle();
                    if (bodyResult instanceof MutableHttpResponse) {
                        return (MutableHttpResponse<?>) bodyResult;
                    } else {
                        response.body(bodyResult);
                    }
                }
                return response;
            });
        } else {
            return chain.proceed(request);
        }
    }

}
```

**Comment by jameskleeh** [comment]
@lightoze Can you test this with 2.0.2?

**Comment by lightoze** [comment]
@jameskleeh The bug is still there - `/test` and `/body` return responses without a cookie, although filter method is invoked.

In the filter switchMap block, the `response.body()` itself is the `Flowable.just(MutableHttpResponse)` returned by `TestController`. I found a relevant piece of code here: https://github.com/micronaut-projects/micronaut-core/blob/e7c5f8c6a7dbc9a7636b70af9ce32bc840a6265d/http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java#L1379-L1407

The problem is that the condition in this block is not satisfied. What is the reasoning behind `finalRoute.isSingleResult() && (finalRoute.isAsync() || Publishers.isSingle(bodyClass))` condition, why just `finalRoute.isSingleResult()` is not enough?

Also there's a similar piece of magic in `buildExecutableRoute` method.

**Comment by lightoze** [comment]
A couple points I made while hacking on this issue:
- there is no reliable way to determine whether reactive publisher will produce one or multiple results at runtime, `Publishers.isSingle` fails on simple cases like `Flowable.just`
- with filters, route signature is not reliable because filters can replace the response, see `ResponseAndStreamSpec` for example

As a result, as little assumptions as possible should be made based on the route when processing a filtered response. At the moment there are places in the code where it is assumed that filtered response reactive/stream/mediaType mode will be the same as declared for the route.

Perhaps this information could be somehow deduced without route info and/or injected into `MutableHttpResponse` before filter chain?

**Comment by yawkat** [comment]
Is this fixed by #4187?

**Comment by yawkat** [comment]
assuming it is fixed
